pkg/sql/importer/import_pgdump.go

```diff
@@ -0,0 +1,581 @@
+			if !ok {
+				return errors.New("cannot get the table name of grant target")
+			}
+			if tableName.ExplicitCatalog {
+				if err := checkCatalogName(tableName.CatalogName, tree.Name(handler.dumpDatabaseName)); err != nil {
+					return err
+				}
+				tableName.CatalogName = TempImportPGDumpDBName
+			}
+		}
+		handler.BufferedDDLStmts = append(handler.BufferedDDLStmts, formatStatement(stmt))
+	case error:
+		if !errors.Is(stmt, errCopyDone) {
+			return stmt
+		}
+	default:
+		return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported %T statement: %s", stmt, stmt))
+	}
+	return nil
+}
+
+// ParseDDLStatementsFromDumpFile parses the DDL statements from the dump file,
+// and replaces all qualified object names to point to the temporary database
+// being imported into.
+// This method returns a PostgresDDLHandler that contains the buffered DDL
+// statements along with other relevant metadata.
+func ParseDDLStatementsFromDumpFile(
+	ctx context.Context,
+	evalCtx *eval.Context,
+	p sql.JobExecContext,
+	dumpFile string,
+	format roachpb.IOFileFormat,
+	maxRowSize int,
+	parentID descpb.ID,
+) (PostgresDDLHandler, error) {
+	handler := PostgresDDLHandler{BufferedDDLStmts: make([]string, 0)}
+	// Open the dump file.
+	store, err := p.ExecCfg().DistSQLSrv.ExternalStorageFromURI(ctx, dumpFile, p.User())
+	if err != nil {
+		return handler, err
+	}
+	defer store.Close()
+
+	raw, err := store.ReadFile(ctx, "")
+	if err != nil {
+		return handler, err
+	}
+	defer raw.Close(ctx)
+	reader, err := decompressingReader(ioctx.ReaderCtxAdapter(ctx, raw), dumpFile, format.Compression)
+	if err != nil {
+		return handler, err
+	}
+	defer reader.Close()
+
+	// Start reading postgres statements.
+	ps := newPostgreStream(ctx, reader, maxRowSize, &unsupportedStmtLogger{} /* unsupportedStmtLogger */)
+	for {
+		stmt, err := ps.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return handler, errors.Wrap(err, "postgres parse error")
+		}
+		if err := bufferDDLPostgresStatement(ctx, evalCtx, stmt.(tree.Statement), p, parentID, &handler); err != nil {
+			return handler, err
+		}
+	}
+	return handler, nil
+}
+
+func runDDLStmtsFromDumpFile(
+	ctx context.Context, stmts []string, txn *kv.Txn, ie sqlutil.InternalExecutor,
+) error {
+	for _, stmt := range stmts {
+		_, err := ie.ExecEx(ctx, "import-pgdump-ddl", txn,
+			sessiondata.InternalExecutorOverride{User: username.NodeUserName(), Database: TempImportPGDumpDBName}, stmt)
+		if err != nil {
+			wrappedErr := errors.Wrapf(err, "executing %s", stmt)
+			fnStmts, parseErr := parser.Parse(stmt)
+			if parseErr != nil {
+				return parseErr
+			}
+			for _, fnStmt := range fnStmts {
+				switch ast := fnStmt.AST.(type) {
+				case *tree.CreateTable, *tree.CreateIndex, *tree.CreateSequence:
+					return errors.WithHintf(wrappedErr, "is the target of the db.target form? "+
+						"please use db.schema.target or schema.target format for target in %q", ast.String())
```

Comment:
Is this hint actionable? I don't know much about pgdump files but I imagine that users can't actually edit them.